### PR TITLE
[java] fix consecutive semicolons problems 

### PR DESCRIPTION
**Pull request by flyinox**
* fix consecutive semicolons problems in packageDeclaration and importDeclaration
* add example

### CODE REVIEW

File: java/java/JavaParser.g4 [hunk 1]
```diff
@@ -39,11 +39,11 @@ compilationUnit
     ;
 
 packageDeclaration
-    : annotation* PACKAGE qualifiedName ';'
+    : annotation* PACKAGE qualifiedName ';'+
```
Comment (msagca) [inline java/java/JavaParser.g4:42]: I don't see such thing in the [spec](https://docs.oracle.com/javase/specs/jls/se20/html/jls-19.html#jls-19-7). Consecutive semicolons usually suggest that there are empty statements, not multiple semicolons after a single statement.

**Review by KvanTTT — CHANGES_REQUESTED** [review]
I agree with @msagca It doesn't look a correct fix.

File: java/java/JavaParser.g4 (hunk 1 above)
Comment (hellozrh) [inline java/java/JavaParser.g4:42]:
```suggestion
    : (annotation* PACKAGE qualifiedName)? ';'
```

File: java/java/JavaParser.g4 [hunk 2]
```diff
@@ -39,11 +39,11 @@ compilationUnit
...
     ;
 
 importDeclaration
-    : IMPORT STATIC? qualifiedName ('.' '*')? ';'
+    : IMPORT STATIC? qualifiedName ('.' '*')? ';'+
```
Comment (hellozrh) [inline java/java/JavaParser.g4:46]:
```suggestion
    : (IMPORT STATIC? qualifiedName ('.' '*')? )?';'
```

**Review by hellozrh — APPROVED** [review]
Your example in the file ConsecutiveSemi.java has Java grammar errors, the ‘package’ statement is not allowed after  'import' statement.

File: java/java/JavaParser.g4 (hunk 1 above)
Comment (flyinox) [reply java/java/JavaParser.g4:42]: apologize for not stating the question clearly. The situation is as follows:
According to the "compileUnit" rule: "packageDeclaration? importDeclaration* typeDeclaration*"
If there are consecutive semicolons after "packageDeclaration" or in a non-last "importDeclaration", these trailing semicolons will be parsed as "typeDeclaration". As a result, the "importDeclaration" and "typeDeclaration" after the trailing semicolons will not have a chance to be correctly parsed.

File: java/java/JavaParser.g4 [hunk 3]
```diff
@@ -39,11 +39,12 @@ compilationUnit
...
 
 packageDeclaration
-    : annotation* PACKAGE qualifiedName ';'+
+    : annotation* PACKAGE qualifiedName ';'
     ;
 
 importDeclaration
-    : IMPORT STATIC? qualifiedName ('.' '*')? ';'+
+    : IMPORT STATIC? qualifiedName ('.' '*')? ';'
+    | ';'
```
Comment (msagca) [inline java/java/JavaParser.g4:47]: This is a better fix even though it makes the rule `compilationUnit` ambiguous (semicolons can be recognized as either `importDeclaration` or `importDeclaration`).

Comment (flyinox) [reply java/java/JavaParser.g4:47]: Yes, and I have noticed this issue in the description of the [spec](https://docs.oracle.com/javase/specs/jls/se20/html/jls-19.html#jls-19-7) as well

**Comment by teverett** [comment]
@KvanTTT  good to merge this?

File: java/java/JavaParser.g4 (hunk 3 above)
Comment (hellozrh) [reply java/java/JavaParser.g4:47]: > 

The demo in the pr show the double semicolon after the statement of packageDeclaration and importDeclaration, but all the other statement with double semicolon ended is correct。So I suggest We should fix other situations。

**Comment by flyinox** [comment]
@KvanTTT I have made some adjustments, please help review and see if it is ready to be merged.

File: java/java/JavaParser.g4 [hunk 4]
```diff
@@ -44,6 +44,7 @@ packageDeclaration
 
 importDeclaration
     : IMPORT STATIC? qualifiedName ('.' '*')? ';'
+    | ';'
```
Comment (KvanTTT) [inline java/java/JavaParser.g4:47]: Please simplify in the following way:

```suggestion
    : (IMPORT STATIC? qualifiedName ('.' '*')?)? ';'
```

**Review by KvanTTT — CHANGES_REQUESTED** [review]

**Comment by KvanTTT** [comment]
Can't we use the following structure?

**Review by KvanTTT — COMMENTED** [review]
Can't we use the following structure instead of your suggestions?

```g4
compilationUnit
    : packageDeclaration? (importDeclaration | ';')* typeDeclaration*
    | moduleDeclaration EOF
    ;

...

importDeclaration
    : IMPORT STATIC? qualifiedName ('.' '*')? ';'
    ;
...

It looks more correct because `importDeclaration` actually always contains `IMPORT` unlike your variant.

File: java/java/JavaParser.g4 (hunk 4 above)
Comment (msagca) [reply java/java/JavaParser.g4:47]: I don't think this makes the rule easier to read/process.

**Comment by msagca** [comment]
As @KvanTTT suggested, empty statements (`;`) inside the rules `importDeclaration` and `typeDeclaration` can be moved to `compilationUnit`:

```antlr
compilationUnit
: packageDeclaration? (importDeclaration | ';')* (typeDeclaration | ';')*
| moduleDeclaration EOF
;
```

**Comment by flyinox** [comment]
> Yes, I also like it more. @flyinox please fix in the suggested way.

Done, `;` have been removed from both the typeDeclaration and importDeclaration

**Comment by teverett** [comment]
@flyinox  thanks!

**Pull request merged**